File: src/M4DSA1/arrays1/continuous_sum_query.py
"""
Time Complexity: O(A+B) | Space Complexity: O(1)
"""

class Solution:
    def solve(self, A, B):
        beggers = [0] * A

        for cherity in B:
            print(cherity)
            beggers[cherity[0]-1] += cherity[2]
            if cherity[1] < A: beggers[cherity[1]] -= cherity[2]

        print(beggers)

        for i in range(len(beggers)):
            beggers[i] += beggers[i-1]

        return beggers
    
solution = Solution()
print(solution.solve(5, [[1,2,10],[2,3,20],[2,5,25]]))

File: src/M4DSA1/arrays1/continuous_sum_query.cpp
#include<iostream>
#include<vector>

using namespace std;

// Time Complexity: O(A+B) | Space Complexity: O(1)

class Soltion {
    public:
    vector<int> solve(int num_of_beggers,vector<vector<int>> const &devotees) {
        vector<int> beggers(num_of_beggers, 0);

        for(auto charity: devotees) {
            beggers[charity[0]-1] += charity[2];
            if(charity[1] < num_of_beggers) 
                beggers[charity[1]] -= charity[2];
        }

        for(int i=1; i<=beggers.size(); i++)
            beggers[i] += beggers[i-1];

        return beggers;
    }
};

int main() {
    Soltion solution;
    vector<int> result = solution.solve(5, {{1, 2, 10}, {2, 3, 20}, {2, 5, 25}});
    
    for (auto money : result)
        cout << money << " ";

}